File: examples/python/01_authentication.py
"""
JETNET API -- Authentication Example

Demonstrates how to:
1. Log in to the JETNET API and obtain tokens
2. Use tokens for subsequent API calls
3. Handle token expiration and refresh

Required environment variables:
    JETNET_EMAIL    -- Your JETNET account email
    JETNET_PASSWORD -- Your JETNET account password
"""

import os
import sys
import time
import requests

BASE = "https://customer.jetnetconnect.com"

TOKEN_TTL = 7 * 3600


def login(email, password):
    """Authenticate and return (bearerToken, apiToken)."""
    r = requests.post(
        f"{BASE}/api/Admin/APILogin",
        json={"emailAddress": email, "password": password},
    )
    r.raise_for_status()
    data = r.json()
    bearer = data.get("bearerToken")
    api_token = data.get("apiToken") or data.get("securityToken")
    if not bearer or not api_token:
        raise RuntimeError("Login succeeded but tokens not found in response.")
    return bearer, api_token


def api(method, path, bearer, token, body=None):
    """Make an authenticated JETNET API call."""
    url = f"{BASE}{path}".replace("{apiToken}", token)
    headers = {
        "Authorization": f"Bearer {bearer}",
        "Content-Type": "application/json",
    }
    r = requests.request(method, url, headers=headers, json=body)
    r.raise_for_status()
    result = r.json()
    status = str(result.get("responsestatus", "")).upper()
    if "ERROR" in status:
        if "INVALID SECURITY TOKEN" in status:
            return None
        raise ValueError(f"JETNET error: {result['responsestatus']}")
    return result


class JetnetClient:
    """Wrapper that manages token lifecycle."""

    def __init__(self, email, password):
        self.email = email
        self.password = password
        self.bearer = None
        self.token = None
        self.token_issued = 0

    def ensure_auth(self):
        if not self.bearer or not self.token or (time.time() - self.token_issued > TOKEN_TTL):
            self.bearer, self.token = login(self.email, self.password)
            self.token_issued = time.time()

    def call(self, method, path, body=None):
        self.ensure_auth()
        result = api(method, path, self.bearer, self.token, body)
        if result is None:
            self.bearer, self.token = login(self.email, self.password)
            self.token_issued = time.time()
            result = api(method, path, self.bearer, self.token, body)
            if result is None:
                raise RuntimeError("Re-login failed; INVALID SECURITY TOKEN persists.")
        return result


def main():
    email = os.environ.get("JETNET_EMAIL")
    password = os.environ.get("JETNET_PASSWORD")
    if not email or not password:
        print("Set JETNET_EMAIL and JETNET_PASSWORD environment variables.")
        sys.exit(1)

    print("Logging in to JETNET API...")
    bearer, api_token = login(email, password)
    print(f"Login successful.")
    print(f"  bearerToken: {bearer[:20]}...")
    print(f"  apiToken:    {api_token[:20]}...")

    print("\nMaking a test call -- looking up tail N1KE...")
    result = api("GET", "/api/Aircraft/getRegNumber/N1KE/{apiToken}", bearer, api_token)
    if result:
        ac = result.get("aircraftresult", {})
        print(f"  Aircraft ID: {ac.get('aircraftid')}")
        print(f"  Make/Model:  {ac.get('make')} {ac.get('model')}")
        print(f"  Serial:      {ac.get('serialnbr')}")
        print(f"  Year:        {ac.get('yearmfr')}")
    else:
        print("  Call returned no result.")

    print("\nUsing JetnetClient wrapper with auto-refresh...")
    client = JetnetClient(email, password)
    result = client.call("GET", "/api/Aircraft/getRegNumber/N1KE/{apiToken}")
    ac = result.get("aircraftresult", {})
    print(f"  Aircraft: {ac.get('make')} {ac.get('model')} (ID: {ac.get('aircraftid')})")


if __name__ == "__main__":
    main()

File: examples/javascript/01_authentication.js
const BASE = process.env.JETNET_BASE_URL || "https://customer.jetnetconnect.com";
const EMAIL = process.env.JETNET_EMAIL;
const PASS = process.env.JETNET_PASSWORD;

if (!EMAIL || !PASS) {
  console.error("Set JETNET_EMAIL and JETNET_PASSWORD environment variables.");
  process.exit(1);
}

async function login() {
  const res = await fetch(`${BASE}/api/Admin/APILogin`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ emailAddress: EMAIL, password: PASS }),
  });

  if (!res.ok) {
    throw new Error(`Login failed: HTTP ${res.status}`);
  }

  const data = await res.json();
  const bearerToken = data.bearerToken;
  const apiToken = data.apiToken || data.securityToken;

  if (!bearerToken || !apiToken) {
    throw new Error("Login succeeded but tokens not found in response.");
  }

  return { bearerToken, apiToken };
}

function isJetnetError(json) {
  return (
    json?.responsestatus &&
    String(json.responsestatus).toUpperCase().includes("ERROR")
  );
}

async function jetnetFetch(bearerToken, apiToken, path, opts = {}) {
  const url = `${BASE}${path}`.replaceAll("{apiToken}", apiToken);
  const res = await fetch(url, {
    ...opts,
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${bearerToken}`,
      ...(opts.headers || {}),
    },
  });

  const text = await res.text();
  let json;
  try {
    json = JSON.parse(text);
  } catch {
    json = { raw: text };
  }

  if (!res.ok || isJetnetError(json)) {
    const msg = isJetnetError(json)
      ? json.responsestatus
      : `HTTP ${res.status}: ${text}`;
    throw new Error(msg);
  }

  return json;
}

async function main() {
  const { bearerToken, apiToken } = await login();
  console.log("Login successful.");
  console.log("Bearer token:", bearerToken.slice(0, 20) + "...");
  console.log("API token:", apiToken);

  const tail = "N123AB";
  const data = await jetnetFetch(
    bearerToken,
    apiToken,
    `/api/Aircraft/getRegNumber/${tail}/{apiToken}`,
    { method: "GET" }
  );

  const ac = data.aircraftresult;
  if (ac) {
    console.log(`\nAircraft found: ${ac.make} ${ac.model} (${ac.regnbr})`);
    console.log(`  Serial: ${ac.serialnbr}`);
    console.log(`  Year: ${ac.yearmfr}`);
    console.log(`  Base: ${ac.baseicao || "N/A"}`);
  } else {
    console.log("No aircraft found for tail:", tail);
  }
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});
